File: src/main/java/com/vc/hard/L214.scala
package com.vc.hard

object L214 {
  def shortestPalindrome(s: String): String = {
    //"aacecaaa#aaacecaa"
    // 01000121012234567

    //"abcd#dcba"
    // 000000001

    val sNew = s + "#" + s.reverse
    val n = sNew.length
    val arr = new Array[Int](n)

    //KMP
    var start = 0
    var end = 1
    while(end < n) {
      if(sNew(end) == sNew(start)) {
        arr(end) = start + 1
        end += 1
        start += 1
      }
      else {
        if(start == 0) {
          arr(end) = 0
          end += 1
        }
        else start = arr(start - 1)
      }
    }
    s.substring(arr(n - 1)).reverse + s
  }
}

File: src/main/java/com/vc/hard/L214.java
package com.vc.hard;

class L214 {
    public String shortestPalindrome(String s) {
        String sNew = s + "#" + reverse(s);

        int n = sNew.length();
        int start = 0;
        int end = 1;
        int[] arr = new int[n];
        while(end < n) {
            if(sNew.charAt(end) == sNew.charAt(start)) {
                arr[end] = start + 1;
                start++;
                end++;
            }
            else {
                if(start == 0) {
                    arr[end] = 0;
                    end++;
                }
                else {
                    start = arr[start - 1];
                }
            }
        }
        // System.out.println(sNew);
        // for(int i = 0; i < arr.length; i++) {
        //     System.out.print(arr[i]);
        // }
        return reverse(s.substring(arr[n - 1])) + s;
    }

    private String reverse(String s) {
        StringBuilder str = new StringBuilder(s);
        return str.reverse().toString();
    }
}
